okhttp/src/main/java/okhttp3/Handshake.kt

```diff
@@ -0,0 +1,156 @@
+import java.security.cert.X509Certificate
+import java.util.ArrayList
+import javax.net.ssl.SSLPeerUnverifiedException
+import javax.net.ssl.SSLSession
+
+/**
+ * A record of a TLS handshake. For HTTPS clients, the client is *local* and the remote server
+ * is its *peer*.
+ *
+ * This value object describes a completed handshake. Use [ConnectionSpec] to set policy
+ * for new handshakes.
+ */
+class Handshake private constructor(
+    private val tlsVersion: TlsVersion,
+    private val cipherSuite: CipherSuite,
+    private val peerCertificates: List<Certificate>,
+    private val localCertificates: List<Certificate>
+) {
+
+  /**
+   * Returns the TLS version used for this connection. This value wasn't tracked prior to OkHttp
+   * 3.0. For responses cached by preceding versions this returns [TlsVersion.SSL_3_0].
+   */
+  fun tlsVersion() = tlsVersion
+
+  /** Returns the cipher suite used for the connection. */
+  fun cipherSuite() = cipherSuite
+
+  /** Returns a possibly-empty list of certificates that identify the remote peer. */
+  fun peerCertificates() = peerCertificates
+
+  /** Returns the remote peer's principle, or null if that peer is anonymous. */
+  fun peerPrincipal(): Principal? {
+    return if (peerCertificates.isNotEmpty())
+      (peerCertificates[0] as X509Certificate).subjectX500Principal
+    else
+      null
+  }
+
+  /** Returns a possibly-empty list of certificates that identify this peer. */
+  fun localCertificates() = localCertificates
+
+  /** Returns the local principle, or null if this peer is anonymous. */
+  fun localPrincipal(): Principal? {
+    return if (localCertificates.isNotEmpty())
+      (localCertificates[0] as X509Certificate).subjectX500Principal
+    else
+      null
+  }
+
+  override fun equals(other: Any?): Boolean {
+    if (other !is Handshake) return false
+    return (tlsVersion == other.tlsVersion
+        && cipherSuite == other.cipherSuite
+        && peerCertificates == other.peerCertificates
+        && localCertificates == other.localCertificates)
+  }
+
+  override fun hashCode(): Int {
+    var result = 17
+    result = 31 * result + tlsVersion.hashCode()
+    result = 31 * result + cipherSuite.hashCode()
+    result = 31 * result + peerCertificates.hashCode()
+    result = 31 * result + localCertificates.hashCode()
+    return result
+  }
+
+  override fun toString(): String {
+    return ("Handshake{"
+        + "tlsVersion="
+        + tlsVersion
+        + " cipherSuite="
+        + cipherSuite
+        + " peerCertificates="
+        + names(peerCertificates)
+        + " localCertificates="
+        + names(localCertificates)
+        + '}'.toString())
+  }
+
+  private fun names(certificates: List<Certificate>): List<String> {
+    val strings = ArrayList<String>()
+
+    for (cert in certificates) {
+      if (cert is X509Certificate) {
+        strings.add(cert.subjectDN.toString())
+      } else {
+        strings.add(cert.type)
+      }
+    }
+    return strings
```

Comment:
Implementation could be
```kotlin
return certificates
    .map {
      if (it is ..) {
        it.subjectDN.toString()
      } else {
        it.type
      }
    }
```
which will also pre-size the new list

Reply:
Nice! 26a75752e3edd3600fbc2f4c0e796134e3b08277